### PR TITLE
Refactoring util functions

### PR DESCRIPTION
**Pull request by tcchiao**
Changes include:
1. a `get_gcm_grid_spec` function to get the definition of the spatial grid for a gcm 
2. change the `open_era5` function to handle multiple variables
3. new file `paths.py` to store all functions that generate paths for caching/storing 
4. a `rechunk_zarr_array_with_caching` function that handles multiple variables and caches

### CODE REVIEW

**Comment by tcchiao** [comment]
@norlandrhagen @orianac a review here would be appreciated, thanks!

File: cmip6_downscaling/data/cmip.py [hunk 1]
```diff
@@ -269,3 +269,19 @@ def gcm_munge(ds: xr.Dataset) -> xr.Dataset:
         ds = ds.drop('height')
     ds = ds.squeeze(drop=True)
     return ds
+
+
+def get_gcm_grid_spec(gcm: str) -> str:
+    gcm_grid = load_cmip(
```
Comment (orianac) [inline cmip6_downscaling/data/cmip.py:276]: this creates such a better string pattern than the ESMF regridder one! i went to look out of curiosity to see if there was anything we were missing and this one is way better. i wonder if we could contribute it as a more informative string-creation utility? https://github.com/pangeo-data/xESMF/blob/25bbb9a001960f46b31e5847c82216d54364ba07/xesmf/frontend.py#L356-L371

File: cmip6_downscaling/workflows/paths.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    Returns
+    -------
+    identifier : str
+        string to be used in obs related paths as specified by the params
+    """
+    if isinstance(variables, str):
+        variables = [variables]
+    var_string = '_'.join(variables)
+    return f'{obs}_{train_period_start}_{train_period_end}_{var_string}_{chunk_by}_chunked'
```
Comment (orianac) [inline cmip6_downscaling/workflows/paths.py]: update these with preferred/clearer chunking variable- i forget which terminology we landed on as being the most intuitive. could also be `chunking_along` but i think we came up with a better solution... 🤔

File: cmip6_downscaling/workflows/paths.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+
+def make_rechunked_obs_path(
```
Comment (orianac) [inline cmip6_downscaling/workflows/paths.py:38]: the only difference among the three functions below is the template at the end, right? could we consolidate them into one function that just takes a template? or reads from templates defined at the top?

File: cmip6_downscaling/workflows/utils.py [hunk 4]
```diff
@@ -65,10 +66,12 @@ def make_rechunker_stores(
         Stores where rechunker will write and the path to the target store
     """
     path_tmp = "az://cmip6/temp/{}.zarr".format(temp_file_name())
-    path_tgt = "az://cmip6/temp/{}.zarr".format(temp_file_name())
     temp_store = fsspec.get_mapper(path_tmp, connection_string=connection_string)
-    target_store = fsspec.get_mapper(path_tgt, connection_string=connection_string)
-    return temp_store, target_store, path_tgt
+
+    if output_path is None:
+        output_path = "az://cmip6/temp/{}.zarr".format(temp_file_name())
```
Comment (orianac) [inline cmip6_downscaling/workflows/utils.py:73]: this is so much cleaner!

File: cmip6_downscaling/workflows/utils.py [hunk 5]
```diff
@@ -346,3 +350,120 @@ def write_dataset(ds: xr.Dataset, path: str, chunks_dims: Tuple = ('time',)) ->
...
+    schema_dict = {}
+    for var in zarr_array.data_vars:
+        schema_dict[var] = DataArraySchema(chunks=chunks_dict[var])
+    target_schema = DatasetSchema(schema_dict)
+
+    # make storage patterns
+    temp_store, target_store, target_path = make_rechunker_stores(connection_string, output_path)
+
+    # check and see if the output is empty, if there is content, check that it's chunked correctly
+    if len(target_store) > 0:
```
Comment (orianac) [inline cmip6_downscaling/workflows/utils.py:419]: when would `target_store` not be >0?

File: cmip6_downscaling/workflows/utils.py [hunk 6]
```diff
@@ -346,3 +350,120 @@ def write_dataset(ds: xr.Dataset, path: str, chunks_dims: Tuple = ('time',)) ->
...
+            else:
+                raise NotImplementedError(
+                    'The content in the output path is incorrectly chunked, but overwrite is disabled.'
+                    'Either clear the output or enable overwrite by setting overwrite=True'
+                )
+
+    # process the input zarr array
+    delete_chunks_encoding(zarr_array)
+    try:
+        # now check if the input is already correctly chunked. If so, save to the output location and return
```
Comment (orianac) [inline cmip6_downscaling/workflows/utils.py:440]: why would we want to do this step? this seems like an unnecessary copy (unless we want to do this just for convenient/organized data storage options)

**Review by orianac — COMMENTED** [review]
just a few questions! and a couple suggestions but none critical

File: cmip6_downscaling/workflows/paths.py (hunk 2 above)
Comment (tcchiao) [reply cmip6_downscaling/workflows/paths.py]: `chunking_approach`!

File: cmip6_downscaling/workflows/paths.py (hunk 3 above)
Comment (tcchiao) [reply cmip6_downscaling/workflows/paths.py:38]: can you say more about how the templates would work?

File: cmip6_downscaling/workflows/utils.py (hunk 5 above)
Comment (tcchiao) [reply cmip6_downscaling/workflows/utils.py:422]: if the target_store is not empty

File: cmip6_downscaling/workflows/utils.py (hunk 6 above)
Comment (tcchiao) [reply cmip6_downscaling/workflows/utils.py:443]: ah, i was thinking then we don't need to rechunk when some other functions are looking for this dataset in this chunking strategy. Maybe we can add an option in the function call `cache` and only saves an extra copy if it's true?

**Pull request merged**